### PR TITLE
Transport detection & Usb serial jtag support

### PR DESCRIPTION
**Pull request by MabezDev**
This PR adds transport detection and a `InputIO` implementation for the USB-SERIAL-JTAG driver.

Currently, draft because the usb serial jtag interrupt is not firing on the esp32s3 see https://github.com/esp-rs/esp-hal/issues/633.

### CODE REVIEW

**Comment by MabezDev** [comment]
https://github.com/esp-rs/esp-hal/pull/664 has resolved the esp32s3 issues. This is now ready for review.

**Comment by MabezDev** [comment]
https://github.com/esp-rs/esp-hal/pull/664 has resolved the esp32s3 issues. This is now ready for review.

File: src/io/usb_serial_jtag.rs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    }
+
+    fn send(&mut self, bytes: &[u8]) {
+        self.write_bytes(bytes).unwrap()
+    }
+}
+
+#[cfg(feature = "esp32c3")]
+#[interrupt]
+unsafe fn USB_SERIAL_JTAG() {
```
Comment (jessebraham) [inline src/io/usb_serial_jtag.rs:23]: I'm fine doing it in another PR, but there interrupt names should be aligned in the PACs then updated here.

**Review by jessebraham — APPROVED** [review]
Left a few nit-picky little comments, up to you whether to address them now or later, but overall LGTM! Thanks for all your work on this!

File: src/io/usb_serial_jtag.rs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+#[cfg(feature = "esp32s3")]
+#[interrupt]
+unsafe fn USB_DEVICE() {
+    isr();
+}
+
+unsafe fn isr() {
+    let usj = crate::hal::peripherals::USB_DEVICE::steal();
```
[the remark this answers is not in the document]
Comment (MabezDev) [reply src/io/usb_serial_jtag.rs:40]: usj was usb serial jtag in my head at the time haha

[the remark this answers is not in the document]
Comment (jessebraham) [reply src/io/usb_serial_jtag.rs:40]: Ahh that make sense and is obvious in hindsight haha, never mind!

File: src/io.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+pub mod uart;
+#[cfg(any(
+    feature = "esp32c3",
+    feature = "esp32s3",
+    feature = "esp32c6",
+    feature = "esp32h2"
+))]
+pub mod usb_serial_jtag;
+
+const RX_QUEUE_SIZE: usize = crate::targets::MAX_WRITE_BLOCK + 0x400;
```
[the remark this answers is not in the document]
Comment (MabezDev) [reply src/io.rs:12]: Sadly I have no idea what that is doing, it was there before this PR. I'm not sure what the `+ 0x400` is for

File: src/io/usb_serial_jtag.rs (hunk 1 above)
Comment (MabezDev) [reply src/io/usb_serial_jtag.rs:23]: I'll do it before merging this, or else I'll forget about it - will also benefit esp-hal.

**Pull request merged**